task-graphs-v2.yaml

```diff
@@ -0,0 +1,370 @@
+      a named argument, depending upon whether `name` is set.
+    type: object
+    properties:
+      name:
+        description: The name of the argument, if present.
+        type: string
+        x-nullable: true
+      value:
+        $ref: '#/definitions/TGArgument'
+
+  TGUDFEnvironment:
+    description: Metadata about the environment where we want to execute a UDF.
+    type: object
+    properties:
+      language:
+        $ref: '#/definitions/UDFLanguage'
+      image_name:
+        description: >
+          The name of the image to use for the execution environment.
+
+
+  #
+  # Parameters
+  #
+
+  TGArgument:
+    description: >
+      An argument provided to a node. This is one of a direct value (i.e.,
+      a raw JSON value) or a `TGSentinel`.
+
+      For example this Python value:
+
+          {"a": [1, "pipe", range(30), None], "b": b"bytes"}
+
+      is encoded thusly (with included comments):
+
+          {  // A dictionary with string keys is JSON-encodable.
+            "a": [  // As is a list.
+              1,
+              "pipe",
+              {"__tiledb_sentinel__": {  // A `range` is replaced with its pickle.
+                "immediate": {
+                  "format": "python_pickle",
+                  "base64_data": "gASVIAAAAAAAAACMCGJ1aWx0aW5zlIwFcmFuZ2WUk5RLAEseSwGHlFKULg=="
+                }
+              }},
+              null
+            ],
+            "b": {"__tiledb_sentinel__": {  // Raw binary data is encoded into base64.
+              "immediate": {
+                "format": "bytes",
+                "base64_data": "Ynl0ZXM="
+              }
+            }}
+          }
+
+
+  # The below messages are not used directly in any API definitions,
+  # but are included as documentation of data encodings.
+
+  TGSentinel:
+    description: >
+      A sentinel value used to replace a value in a call that is made.
+      This will be replaced with the actual value to be sent to the UDF
+      environment, either by the server or the client.
+    type: object
+    properties:
+      __tiledb_sentinel__:
+        $ref: '#/definitions/TGSentinelValue'
+  TGSentinelValue:
+    description: >
+      The key that indicates that this value is not to be passed directly to
+      a UDF, but is instead some kind of upstream node or other special value.
+      Exactly one of these keys should be set.
+    type: object
+    properties:
+      node_output:
+        type: string
+        x-nullable: true
+        description: Substitutes the value of this node at call time.
+      array:
+        $ref: '#/definitions/TGArrayNodeData'
+      immediate:
+        $ref: '#/definitions/TGImmediateValue'
+      __escape__:
+        x-nullable: true
+        description: >
+          “Escape sequence” for if the user has a dictionary with
+          a `__tiledb_sentinel__` key.
+
+  TGImmediateValue:
+    description: >
+      An immediate value (i.e., not a node output) to be passed to an argument
```

Comment:
All data passing is via `TGImmediateValue` i can't see how a json value can be passed directly to a node. The use of this type suggests that even json is base64 encoded